### DISCUSSION TITLE
[Customization guide] How to make ← and → in menu mean to leave the completion mode and edit the command line?

### DISCUSSION THREAD

**Discussion by ImpleLee** [marlonrichert/zsh-autocomplete#630] [Q&A] [answered]
## What do you want?
Add an customization guide in the README to describe how to make ← and → in menu mean editing the current command line.

## Why do you want this?
I think it is straightforward to make ← and → mean editing instead of menu selection.

## Who else would benefit from this?
Others that think ← and → is more suited as editing instead of menu selection.
This issue only wants to add an entry in the README, so this does not change current behavior.

## How should it work?
I actually don't know. That is part of the reason that I want to make a feature request.

Given the following situation:
* I am in the completion menu.

When I perform the following steps:
1. I select some completion by ↑ and ↓.
2. Then I hit ← or →.

Then I expect the following to happen:
* I can edit the current command line (the completion I accepted).

**Comment by marlonrichert** [comment, marked as answer]
```zsh
bindkey -M menuselect '<escape code for ←>' .backward-char
bindkey -M menuselect '<escape code for →>'  .forward-char
```
Documented in
* [Zsh Modules > Menu selection](https://zsh.sourceforge.io/Doc/Release/Zsh-Modules.html#Menu-selection)
* [Zsh Line Editor > Movement](https://zsh.sourceforge.io/Doc/Release/Zsh-Line-Editor.html#Movement)

**Reply by julien-blanchon** [reply]
What are the escapes right key, this don't work
```
bindkey -M menuselect '\e[D' backward-char
bindkey -M menuselect '\e[C' forward-char
```

**Reply by julien-blanchon** [reply]
Also 
```
bindkey -M menuselect '^[[D' backward-char
bindkey -M menuselect '^[[C' forward-char
```

Don't work

**Reply by JerryZhongJ** [reply]
For me it doesn't work either. I added these to the end of `.zshrc`:
```
bindkey -M menuselect '^[[D' .backward-char
bindkey -M menuselect '^[[C' .forward-char
```

**Reply by amomchilov** [reply]
Oh boy, it took me like an hour to figure this out, but this fails when cusor keys mode ("DECCKM") is enabled. The sequnces are `^[OD` and `^[OC` instead.

That's what the `prefix=( '\e'{\[,O} )` is for [in the script from the `README.md`](https://github.com/marlonrichert/zsh-autocomplete/blob/c7b65508fd3a016dc9cdb410af9ee7806b3f9be1/README.md#reset--and-). It gets expanded out to call `bindkey` for both `^[[` and `^[O`.

**Reply by amomchilov** [reply]
I'll add, I also figured out some more combinations:

| Keyboard shortcut | Characters sent | Effect |
|-------------------|-----------------|--------|
| <kbd>Option ⌥</kbd> + <kbd>←</kbd> | `^[b` | move left by a word
| <kbd>Option ⌥</kbd> + <kbd>→</kbd>  | `^[f` | move right by a word
| <kbd>Command ⌘</kbd> + <kbd>←</kbd> | `^A` | move to the start of the line
| <kbd>Command ⌘</kbd> + <kbd>→</kbd> | `^E` | move to the end of the line

```zsh
bindkey -M menuselect "^[OD"   .backward-char # ←
bindkey -M menuselect "^[OC"    .forward-char # →
bindkey -M menuselect "^[b"    .backward-word # ⌥ + ←
bindkey -M menuselect "^[f"     .forward-word # ⌥ + →
bindkey -M menuselect "^A" .beginning-of-line # ⌘ + ←
bindkey -M menuselect "^E"       .end-of-line # ⌘ + →
```

The 4th and 6th ones are kind of redundant, because your insertion points already starts out on the end of the word/line. I just added it for completeness.

**Comment by JerryZhongJ** [comment]
I borrowed the idea from `README.md`, and this one work for me:
```
() {
   local -a prefix=( '\e'{\[,O} )
   local -a left=( ${^prefix}D ) right=( ${^prefix}C )
   local key=
   for key in $left[@]; do
      bindkey -M menuselect "$key" .backward-char
   done
   for key in $right[@]; do
      bindkey -M menuselect "$key" .forward-char
   done
}
```

**Comment by ImpleLee** [comment]
I personally use the following snippet and it works fine.
```zsh
bindkey -M menuselect "$terminfo[kcub1]" .backward-char
bindkey -M menuselect "$terminfo[kcuf1]" .forward-char      
```
